examples/keras_recipes/memory_efficient_embeddings.py

```diff
@@ -0,0 +1,419 @@
+ZipFile("movielens.zip", "r").extractall()
+
+ratings_data = pd.read_csv(
+    "ml-1m/ratings.dat",
+    sep="::",
+    names=["user_id", "movie_id", "rating", "unix_timestamp"],
+)
+
+ratings_data["movie_id"] = ratings_data["movie_id"].apply(lambda x: f"movie_{x}")
+ratings_data["user_id"] = ratings_data["user_id"].apply(lambda x: f"user_{x}")
+ratings_data["rating"] = ratings_data["rating"].apply(lambda x: float(x))
+del ratings_data["unix_timestamp"]
+
+print(f"Number of users: {len(ratings_data.user_id.unique())}")
+print(f"Number of movies: {len(ratings_data.movie_id.unique())}")
+print(f"Number of ratings: {len(ratings_data.index)}")
+
+"""
+## Create train and eval data splits
+"""
+
+random_selection = np.random.rand(len(ratings_data.index)) <= 0.85
+train_data = ratings_data[random_selection]
+eval_data = ratings_data[~random_selection]
+
+train_data.to_csv("train_data.csv", index=False, sep="|", header=False)
+eval_data.to_csv("eval_data.csv", index=False, sep="|", header=False)
+print(f"Train data split: {len(train_data.index)}")
+print(f"Eval data split: {len(eval_data.index)}")
+print("Train and eval data files are saved.")
+
+"""
+## Define dataset metadata and hyperparameters
+"""
+
+csv_header = list(ratings_data.columns)
+user_vocabulary = list(ratings_data.user_id.unique())
+movie_vocabulary = list(ratings_data.movie_id.unique())
+target_feature_name = "rating"
+learning_rate = 0.001
+batch_size = 128
+num_epochs = 3
+base_embedding_dim = 64
+
+"""
+## Train and evaluate the model
+"""
+
+
+def get_dataset_from_csv(csv_file_path, batch_size=128, shuffle=True):
+    return tf.data.experimental.make_csv_dataset(
+        csv_file_path,
+        batch_size=batch_size,
+        column_names=csv_header,
+        label_name=target_feature_name,
+        num_epochs=1,
+        header=False,
+        field_delim="|",
+        shuffle=shuffle,
+    )
+
+
+def run_experiment(model):
+    # Compile the model.
+    model.compile(
+        optimizer=keras.optimizers.Adam(learning_rate),
+        loss=tf.keras.losses.MeanSquaredError(),
+        metrics=[keras.metrics.MeanAbsoluteError(name="mae")],
+    )
+    # Read the training data.
+    train_dataset = get_dataset_from_csv("train_data.csv", batch_size)
+    # Read the test data.
+    eval_dataset = get_dataset_from_csv("eval_data.csv", batch_size, shuffle=False)
+    # Fit the model with the training data.
+    history = model.fit(train_dataset, epochs=num_epochs, validation_data=eval_dataset)
+
+    # Evaluate the model on the test data.
+    _, rmse = model.evaluate(eval_dataset, verbose=0)
+    print(f"Eval MAE: {round(rmse, 3)}")
+    return history
+
+
+"""
+## Experiment 1: standard collaborative filtering model
+"""
+
+"""
+### Implement embedding encoder
+"""
+
+
+def embedding_encoder(vocabulary, embedding_dim, num_oov_indices=0, name=None):
+    return keras.Sequential(
+        [
+            StringLookup(
```

Comment:
On GPU/TPU you will want to put the StringLookup layer in the tf.data pipeline, not in the model (and only migrate it back to the model to export an inference model). Though on CPU having StringLookup in the model is fine

Reply:
To be clear, I'm not sure whether this matters in this case. I'd recommend benchmarking the two ways on the Colab GPU. If putting the layer in the tf.data pipeline doesn't provide a speedup, then let's keep it in the model.

Reply:
Ok, I tried it on a GPU Colab and putting StringLookup in the tf.data pipeline didn't provide a significant speendup (1ms difference per step!).

Maybe with a bit of optimization of the tf.data pipeline it could, however I would keep the model as is.